### PR TITLE
Patch Vite and update dependencies

### LINKED ISSUE #96: build error in SWR 2.2.0
```tsx
"use client";

import { useState } from "react";
import useSWR from 'swr'

export const Counter = () => {
  // @ts-ignore
  const { data } = useSWR('hello', () => 'world', {
    suspense: true
  })
  const [count, setCount] = useState(0);
  return (
    <div style={{ border: "3px blue dashed", margin: "1em", padding: "1em" }}>
      <p>Count: {count}</p>
      <button onClick={() => setCount((c) => c + 1)}>Increment</button>
      <h3>This is a client component.</h3>
      <div>
        {data}
      </div>
    </div>
  );
};
```

```shell
/Users/himself65/.nvm/versions/node/v18.16.1/bin/npm run build

> affine-waku@0.1.0 build
> waku build

vite v4.3.9 building SSR bundle for production...
✓ 6 modules transformed.
✓ built in 42ms
"default" is not exported by "node_modules/swr/core/dist/react-server.mjs", imported by "src/Counter.tsx".
file: /Users/himself65/Code/affine-waku/src/Counter.tsx:4:7
2: 
3: import { useState } from "react";
4: import useSWR from 'swr'
          ^
5: 
6: export const Counter = () => {
file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:2124
        base = Object.assign(new Error(base.message), base);
                             ^

Error [RollupError]: "default" is not exported by "node_modules/swr/core/dist/react-server.mjs", imported by "src/Counter.tsx".
    at error (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:2124:30)
    at Module.error (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:13463:16)
    at Module.traceVariable (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:13884:29)
    at ModuleScope.findVariable (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:12429:39)
    at ReturnValueScope.findVariable (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:6965:38)
    at ChildScope.findVariable (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:6965:38)
    at Identifier.bind (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:8127:40)
    at CallExpression.bind (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:5738:23)
    at CallExpression.bind (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:9680:15)
    at VariableDeclarator.bind (file:///Users/himself65/Code/affine-waku/node_modules/rollup/dist/es/shared/node-entry.js:5738:23) {
  binding: 'default',
  code: 'MISSING_EXPORT',
  exporter: '/Users/himself65/Code/affine-waku/node_modules/swr/core/dist/react-server.mjs',
  id: '/Users/himself65/Code/affine-waku/src/Counter.tsx',
  url: 'https://rollupjs.org/troubleshooting/#error-name-is-not-exported-by-module',
  pos: 103,
  loc: {
    column: 7,
    file: '/Users/himself65/Code/affine-waku/src/Counter.tsx',
    line: 4
  },
  frame: '2: \n' +
    '3: import { useState } from "react";\n' +
    "4: import useSWR from 'swr'\n" +
    '          ^\n' +
    '5: \n' +
    '6: export const Counter = () => {',
  watchFiles: [
    '/Users/himself65/Code/affine-waku/entries.ts',
    '/Users/himself65/Code/affine-waku/package.json',
    '/Users/himself65/Code/affine-waku/node_modules/waku/dist/server.js',
    '/Users/himself65/Code/affine-waku/src/App.tsx',
    '/Users/himself65/Code/affine-waku/src/Counter.tsx',
    '/Users/himself65/Code/affine-waku/node_modules/swr/core/dist/react-server.mjs',
    '/Users/himself65/Code/affine-waku/node_modules/swr/_internal/dist/react-server.mjs'
  ]
}

Node.js v18.16.1

Process finished with exit code 1
```

Upstream: https://github.com/vercel/swr/issues/2700

---

repo: https://github.com/himself65/affine-waku/blob/issue-swr/src/Counter.tsx

```shell
yarn install
yarn build
```

### LINKED ISSUE #111: Error with latest react builds
`18.3.0-canary-e91142dd6-20230705` works, but `18.3.0-canary-fdc8c81e0-20230707` fails.

> [vite] Error when evaluating SSR module /src/components/App.tsx: failed to import "/src/components/Counter.tsx"
|- Error: The React Server Writer cannot be used outside a react-server environment. You must configure Node.js using the `--conditions react-server` flag.

It seems like https://github.com/facebook/react/pull/27033 changes to use `import {registerServerReference} from "react-server-dom-webpack/server";` in the transformed file with packages/react-server-dom-webpack/src/ReactFlightWebpackNodeLoader.js, which might cause the error.

Not yet sure how to fix it.

### LINKED ISSUE #114: hooks cannot be used in server component
I try to use hooks in server component, but it reported that 「Cannot read properties of null (reading 'useContext')」

This bug is also similar with https://github.com/dai-shi/waku/pull/103#issuecomment-1633861236 since react-tweet use 「useMemo」 in the server component

codesandbox: https://codesandbox.io/p/sandbox/waku-example-counter-forked-4rg8fz?file=%2Fsrc%2Fcomponents%2FApp.tsx%3A7%2C38

I figured it out the reason, but I do not know how to fix it.

vite.ssrLoadModule will resolve package with override conditions https://github.com/vitejs/vite/blob/main/packages/vite/src/node/ssr/ssrModuleLoader.ts#L132. So when the user code bundled in ssrLoadModule, the react package will be resolved as index.js instead of react.shared-subset.js because of lacking react-server condition.

And the react-dom-server package resolve react as react.shared-subset.js, so there will appear two React instance.

After I forcely changed the vite code by adding react-server condition, it worked.
CleanShot 2023-08-01 at 00 03 47@2x

Obviously it is not a correct way to solve this problem and I have no idea how to fix it, so I leave a message about what I found here.

### PR DESCRIPTION
**Pull request by dai-shi**
:warning: This would only work on Linux and MacOS for now.

Test with:
- [x] npm
- [x] yarn1
- pnpm -> will test after releasing

Close #96
Close #111 
Close #114

### CODE REVIEW

File: examples/02_async/package.json [hunk 1]
```diff
@@ -10,14 +10,19 @@
...
   "devDependencies": {
-    "@types/react": "^18.2.21",
-    "@types/react-dom": "^18.2.7",
+    "@types/react": "^18.2.24",
+    "@types/react-dom": "^18.2.8",
     "typescript": "^5.2.2"
+  },
+  "pnpm": {
+    "patchedDependencies": {
+      "vite@4.4.10": "patches/vite@4.4.10.patch"
```
Comment (himself65) [inline examples/02_async/package.json:25]: Seems like you didn't upload the patch

**Review by himself65 — COMMENTED** [review]
I think the beat solution is waiting for upstream give a patch version. Since waku is a library, we cannot control let user to install the patch automatically.

File: packages/waku/package.json [hunk 2]
```diff
@@ -51,33 +51,35 @@
...
     "node": ">=18.0.0"
   },
   "dependencies": {
-    "@swc/core": "1.3.83",
-    "@vitejs/plugin-react": "^4.0.4",
+    "@swc/core": "1.3.91",
+    "@vitejs/plugin-react": "^4.1.0",
     "busboy": "^1.6.0",
-    "vite": "^4.4.9"
+    "vite": "4.4.10"
```
Comment (himself65) [inline packages/waku/package.json:72]: Also, doesn't `vite` should be in the peer dependencies?

**Comment by dai-shi** [comment]
> I think the beat solution is waiting for upstream give a patch version.

We waited for several months and it seems to take more time.
This has stopped us developing/testing things. Rather than stopping, we should move forward.

> Since waku is a library, we cannot control let user to install the patch automatically.

I would actually like to establish the patch solution with libraries. So, happy to learn the cases where the patch isn't applied automatically. For now, pnpm is challenging and if users don't use `create-waku` it fails. I wonder if there's a better idea.

File: packages/waku/package.json (hunk 2 above)
Comment (dai-shi) [reply packages/waku/package.json:72]: Thanks. I don't think we need it in the peer deps as it's already in the deps.

**Comment by dai-shi** [comment]
@Tobbe said they also patched it: https://github.com/vitejs/vite/pull/13487#issuecomment-1695295029

And, I just found https://github.com/vitejs/vite/pull/14498

**Comment by dai-shi** [comment]
Let's merge this and wait a new release from Vite.

**Pull request merged**